File: swift/src/Pathspec.swift
/// MuonGit - Pathspec pattern matching
/// Parity: libgit2 src/libgit2/pathspec.c
import Foundation

/// Flags for pathspec matching behavior
public struct PathspecFlags: Sendable {
    public var ignoreCase: Bool
    public var noGlob: Bool
    public var noMatchError: Bool
    public var findFailures: Bool

    public init(ignoreCase: Bool = false, noGlob: Bool = false, noMatchError: Bool = false, findFailures: Bool = false) {
        self.ignoreCase = ignoreCase
        self.noGlob = noGlob
        self.noMatchError = noMatchError
        self.findFailures = findFailures
    }
}

/// A single pathspec pattern
private struct PathspecPattern {
    let pattern: String
    let negated: Bool
    let hasWildcard: Bool
    let matchAll: Bool
}

/// Compiled pathspec for matching file paths
public struct Pathspec: Sendable {
    private let patterns: [PathspecPattern]

    /// Create a new pathspec from a list of patterns
    public init(patterns: [String]) {
        self.patterns = patterns.map { Self.parsePattern($0) }
    }

    /// Check if a path matches this pathspec
    public func matchesPath(_ path: String, flags: PathspecFlags = PathspecFlags()) -> Bool {
        if patterns.isEmpty { return true }

        var matched = false
        for pattern in patterns {
            if pattern.matchAll {
                matched = !pattern.negated
                continue
            }

            let doesMatch: Bool
            if flags.noGlob {
                doesMatch = pathMatchesLiteral(path: path, pattern: pattern.pattern, ignoreCase: flags.ignoreCase)
            } else {
                doesMatch = pathMatchesGlob(path: path, pattern: pattern.pattern, ignoreCase: flags.ignoreCase)
            }

            if doesMatch {
                matched = !pattern.negated
            }
        }
        return matched
    }

    /// Match this pathspec against a list of paths
    public func matchPaths(_ paths: [String], flags: PathspecFlags = PathspecFlags()) -> PathspecMatchResult {
        var matches: [String] = []
        var matchedPatterns = [Bool](repeating: false, count: patterns.count)

        for path in paths {
            if matchesPath(path, flags: flags) {
                matches.append(path)
            }
            for (i, pattern) in patterns.enumerated() {
                if !pattern.negated && patternMatchesPath(path: path, pattern: pattern, flags: flags) {
                    matchedPatterns[i] = true
                }
            }
        }

        var failures: [String] = []
        if flags.findFailures {
            for (i, pattern) in patterns.enumerated() {
                if !pattern.negated && !matchedPatterns[i] {
                    failures.append(pattern.pattern)
                }
            }
        }

        return PathspecMatchResult(matches: matches, failures: failures)
    }

    public var count: Int { patterns.count }
    public var isEmpty: Bool { patterns.isEmpty }

    // MARK: - Parsing

    private static func parsePattern(_ pat: String) -> PathspecPattern {
        var pattern = pat
        var negated = false

        if pattern.hasPrefix("!") {
            negated = true
            pattern = String(pattern.dropFirst())
        } else if pattern.hasPrefix("\\!") {
            pattern = String(pattern.dropFirst())
        }

        if pattern.hasPrefix("/") {
            pattern = String(pattern.dropFirst())
        }

        let matchAll = pattern == "*" || pattern.isEmpty
        let hasWildcard = pattern.contains("*") || pattern.contains("?") || pattern.contains("[")

        return PathspecPattern(pattern: pattern, negated: negated, hasWildcard: hasWildcard, matchAll: matchAll)
    }
}

/// Result of matching a pathspec against a list of paths
public struct PathspecMatchResult: Sendable {
    public let matches: [String]
    public let failures: [String]
}

// MARK: - Internal matching

private func patternMatchesPath(path: String, pattern: PathspecPattern, flags: PathspecFlags) -> Bool {
    if pattern.matchAll { return true }
    if flags.noGlob {
        return pathMatchesLiteral(path: path, pattern: pattern.pattern, ignoreCase: flags.ignoreCase)
    }
    return pathMatchesGlob(path: path, pattern: pattern.pattern, ignoreCase: flags.ignoreCase)
}

private func pathMatchesLiteral(path: String, pattern: String, ignoreCase: Bool) -> Bool {
    let p = ignoreCase ? pattern.lowercased() : pattern
    let t = ignoreCase ? path.lowercased() : path

    if t == p { return true }
    if t.hasPrefix(p) && t.dropFirst(p.count).first == "/" { return true }
    return false
}

private func pathMatchesGlob(path: String, pattern: String, ignoreCase: Bool) -> Bool {
    let p = ignoreCase ? pattern.lowercased() : pattern
    let t = ignoreCase ? path.lowercased() : path

    // Handle ** (any number of path levels)
    if p.hasPrefix("**/") {
        let sub = String(p.dropFirst(3))
        if wildmatch(pattern: sub, text: t) { return true }
        var remaining = t
        while let idx = remaining.firstIndex(of: "/") {
            remaining = String(remaining[remaining.index(after: idx)...])
            if wildmatch(pattern: sub, text: remaining) { return true }
        }
        return false
    }

    // If pattern has no '/', match against basename
    if !p.contains("/") {
        let basename = t.split(separator: "/").last.map(String.init) ?? t
        if wildmatch(pattern: p, text: basename) { return true }
    }

    // Full path match
    if wildmatch(pattern: p, text: t) { return true }

    // Directory prefix
    let stripped = p.hasSuffix("/") ? String(p.dropLast()) : p
    if t.hasPrefix(stripped) && t.dropFirst(stripped.count).first == "/" { return true }

    return false
}

private func wildmatch(pattern: String, text: String) -> Bool {
    let p = Array(pattern)
    let t = Array(text)
    return wildmatchInner(p, 0, t, 0)
}

private func wildmatchInner(_ pattern: [Character], _ pi: Int, _ text: [Character], _ ti: Int) -> Bool {
    if pi == pattern.count && ti == text.count { return true }
    if pi == pattern.count { return false }

    if pattern[pi] == "*" {
        // * matches everything except /
        if wildmatchInner(pattern, pi + 1, text, ti) { return true }
        if ti < text.count && text[ti] != "/" {
            return wildmatchInner(pattern, pi, text, ti + 1)
        }
        return false
    }

    if ti >= text.count { return false }

    if pattern[pi] == "?" {
        return text[ti] != "/" && wildmatchInner(pattern, pi + 1, text, ti + 1)
    }

    if pattern[pi] == "[" {
        if let (matched, nextPi) = matchCharClass(pattern, pi + 1, text[ti]) {
            if matched { return wildmatchInner(pattern, nextPi, text, ti + 1) }
        }
        return false
    }

    if pattern[pi] == text[ti] {
        return wildmatchInner(pattern, pi + 1, text, ti + 1)
    }

    return false
}

private func matchCharClass(_ pattern: [Character], _ start: Int, _ ch: Character) -> (Bool, Int)? {
    var i = start
    var negated = false
    if i < pattern.count && pattern[i] == "!" {
        negated = true
        i += 1
    }

    var matched = false
    while i < pattern.count && pattern[i] != "]" {
        if i + 2 < pattern.count && pattern[i + 1] == "-" {
            if ch >= pattern[i] && ch <= pattern[i + 2] { matched = true }
            i += 3
        } else {
            if ch == pattern[i] { matched = true }
            i += 1
        }
    }

    guard i < pattern.count && pattern[i] == "]" else { return nil }
    let result = negated ? !matched : matched
    return (result, i + 1)
}

File: rust/src/pathspec.rs
//! Pathspec pattern matching
//! Parity: libgit2 src/libgit2/pathspec.c

/// Flags for pathspec matching behavior
#[derive(Debug, Clone, Copy, Default)]
pub struct PathspecFlags {
    pub ignore_case: bool,
    pub no_glob: bool,
    pub no_match_error: bool,
    pub find_failures: bool,
}

/// A single pathspec pattern
#[derive(Debug, Clone)]
struct PathspecPattern {
    pattern: String,
    negated: bool,
    match_all: bool,
}

/// Compiled pathspec for matching file paths
#[derive(Debug, Clone)]
pub struct Pathspec {
    patterns: Vec<PathspecPattern>,
}

/// Result of matching a pathspec against a list of paths
#[derive(Debug, Clone)]
pub struct PathspecMatchResult {
    pub matches: Vec<String>,
    pub failures: Vec<String>,
}

impl Pathspec {
    /// Create a new pathspec from a list of patterns
    pub fn new(patterns: &[&str]) -> Self {
        let mut compiled = Vec::new();
        for &pat_str in patterns {
            compiled.push(parse_pattern(pat_str));
        }
        Pathspec { patterns: compiled }
    }

    /// Check if a path matches this pathspec
    pub fn matches_path(&self, path: &str, flags: &PathspecFlags) -> bool {
        if self.patterns.is_empty() {
            return true;
        }

        let mut matched = false;

        for pattern in &self.patterns {
            if pattern.match_all {
                matched = !pattern.negated;
                continue;
            }

            let does_match = if flags.no_glob {
                path_matches_literal(path, &pattern.pattern, flags.ignore_case)
            } else {
                path_matches_glob(path, &pattern.pattern, flags.ignore_case)
            };

            if does_match {
                matched = !pattern.negated;
            }
        }

        matched
    }

    /// Match this pathspec against a list of paths
    pub fn match_paths(&self, paths: &[&str], flags: &PathspecFlags) -> PathspecMatchResult {
        let mut matches = Vec::new();
        let mut matched_patterns = vec![false; self.patterns.len()];

        for &path in paths {
            if self.matches_path(path, flags) {
                matches.push(path.to_string());
            }
            // Track which patterns matched for failure detection
            for (i, pattern) in self.patterns.iter().enumerate() {
                if !pattern.negated && pattern_matches_path(path, pattern, flags) {
                    matched_patterns[i] = true;
                }
            }
        }

        let failures = if flags.find_failures {
            self.patterns
                .iter()
                .enumerate()
                .filter(|(i, p)| !p.negated && !matched_patterns[*i])
                .map(|(_, p)| p.pattern.clone())
                .collect()
        } else {
            Vec::new()
        };

        PathspecMatchResult { matches, failures }
    }

    /// Number of patterns
    pub fn len(&self) -> usize {
        self.patterns.len()
    }

    /// Whether the pathspec is empty
    pub fn is_empty(&self) -> bool {
        self.patterns.is_empty()
    }
}

/// Parse a single pattern string
fn parse_pattern(pat: &str) -> PathspecPattern {
    let mut pattern = pat.to_string();
    let mut negated = false;

    // Handle negation
    if pattern.starts_with('!') {
        negated = true;
        pattern = pattern[1..].to_string();
    } else if pattern.starts_with('\\') && pattern.len() > 1 && pattern.as_bytes()[1] == b'!' {
        // Escaped negation
        pattern = pattern[1..].to_string();
    }

    // Strip leading slash (anchors to root)
    if pattern.starts_with('/') {
        pattern = pattern[1..].to_string();
    }

    let match_all = pattern == "*" || pattern.is_empty();
    PathspecPattern {
        pattern,
        negated,
        match_all,
    }
}

fn pattern_matches_path(path: &str, pattern: &PathspecPattern, flags: &PathspecFlags) -> bool {
    if pattern.match_all {
        return true;
    }
    if flags.no_glob {
        path_matches_literal(path, &pattern.pattern, flags.ignore_case)
    } else {
        path_matches_glob(path, &pattern.pattern, flags.ignore_case)
    }
}

/// Literal path matching (prefix + exact)
fn path_matches_literal(path: &str, pattern: &str, ignore_case: bool) -> bool {
    let (p, t) = if ignore_case {
        (pattern.to_lowercase(), path.to_lowercase())
    } else {
        (pattern.to_string(), path.to_string())
    };

    // Exact match
    if t == p {
        return true;
    }

    // Directory prefix match: pattern "dir" matches "dir/file"
    if t.starts_with(&p) && t.as_bytes().get(p.len()) == Some(&b'/') {
        return true;
    }

    false
}

/// Glob-based path matching
fn path_matches_glob(path: &str, pattern: &str, ignore_case: bool) -> bool {
    let (p, t) = if ignore_case {
        (pattern.to_lowercase(), path.to_lowercase())
    } else {
        (pattern.to_string(), path.to_string())
    };

    // Handle ** (any number of path levels)
    if let Some(sub) = p.strip_prefix("**/") {
        // Match at any directory level
        if wildmatch(sub, &t) {
            return true;
        }
        // Also try at every directory level
        let mut pos = 0;
        while let Some(idx) = t[pos..].find('/') {
            if wildmatch(sub, &t[pos + idx + 1..]) {
                return true;
            }
            pos += idx + 1;
        }
        return false;
    }

    // If pattern has no '/', match against basename only (git behavior)
    if !p.contains('/') {
        let basename = t.rsplit('/').next().unwrap_or(&t);
        if wildmatch(&p, basename) {
            return true;
        }
    }

    // Standard glob match against full path
    if wildmatch(&p, &t) {
        return true;
    }

    // Trailing slash stripped: pattern "dir" matches "dir/file"
    let stripped = p.trim_end_matches('/');
    if t.starts_with(stripped) && t.as_bytes().get(stripped.len()) == Some(&b'/') {
        return true;
    }

    false
}

/// Wildcard matching (supports *, ?, and [...])
fn wildmatch(pattern: &str, text: &str) -> bool {
    let p: Vec<char> = pattern.chars().collect();
    let t: Vec<char> = text.chars().collect();
    wildmatch_inner(&p, &t)
}

fn wildmatch_inner(pattern: &[char], text: &[char]) -> bool {
    match (pattern.first(), text.first()) {
        (None, None) => true,
        (None, Some(_)) => false,
        (Some(&'*'), _) => {
            // * matches everything except /
            // Try matching zero characters
            if wildmatch_inner(&pattern[1..], text) {
                return true;
            }
            // Try matching one or more characters (but not /)
            if let Some(&ch) = text.first() {
                if ch != '/' {
                    return wildmatch_inner(pattern, &text[1..]);
                }
            }
            false
        }
        (Some(&'?'), Some(&ch)) => {
            if ch != '/' {
                wildmatch_inner(&pattern[1..], &text[1..])
            } else {
                false
            }
        }
        (Some(&'['), _) => {
            // Character class
            if let Some((matched, rest_pattern)) = match_char_class(&pattern[1..], text.first()) {
                if matched {
                    return wildmatch_inner(rest_pattern, &text[1..]);
                }
            }
            false
        }
        (Some(&a), Some(&b)) if a == b => wildmatch_inner(&pattern[1..], &text[1..]),
        _ => false,
    }
}

/// Match a character class [abc] or [!abc] or [a-z]
fn match_char_class<'a>(pattern: &'a [char], ch: Option<&char>) -> Option<(bool, &'a [char])> {
    let ch = ch?;
    let mut negated = false;
    let mut i = 0;

    if i < pattern.len() && pattern[i] == '!' {
        negated = true;
        i += 1;
    }

    let mut matched = false;
    while i < pattern.len() && pattern[i] != ']' {
        if i + 2 < pattern.len() && pattern[i + 1] == '-' {
            // Range
            if *ch >= pattern[i] && *ch <= pattern[i + 2] {
                matched = true;
            }
            i += 3;
        } else {
            if *ch == pattern[i] {
                matched = true;
            }
            i += 1;
        }
    }

    if i < pattern.len() && pattern[i] == ']' {
        let result = if negated { !matched } else { matched };
        Some((result, &pattern[i + 1..]))
    } else {
        None // Malformed character class
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_matching() {
        let ps = Pathspec::new(&["*.rs"]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("src/main.rs", &flags));
        assert!(!ps.matches_path("src/main.py", &flags));
    }

    #[test]
    fn test_directory_prefix() {
        let ps = Pathspec::new(&["src"]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("src", &flags));
        assert!(ps.matches_path("src/main.rs", &flags));
        assert!(!ps.matches_path("test/main.rs", &flags));
    }

    #[test]
    fn test_negation() {
        let ps = Pathspec::new(&["*.rs", "!test_*.rs"]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("main.rs", &flags));
        assert!(!ps.matches_path("test_main.rs", &flags));
    }

    #[test]
    fn test_double_star() {
        let ps = Pathspec::new(&["**/test.rs"]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("test.rs", &flags));
        assert!(ps.matches_path("src/test.rs", &flags));
        assert!(ps.matches_path("a/b/c/test.rs", &flags));
        assert!(!ps.matches_path("test.py", &flags));
    }

    #[test]
    fn test_question_mark() {
        let ps = Pathspec::new(&["?.rs"]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("a.rs", &flags));
        assert!(!ps.matches_path("ab.rs", &flags));
    }

    #[test]
    fn test_char_class() {
        let ps = Pathspec::new(&["[abc].rs"]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("a.rs", &flags));
        assert!(ps.matches_path("b.rs", &flags));
        assert!(!ps.matches_path("d.rs", &flags));
    }

    #[test]
    fn test_ignore_case() {
        let ps = Pathspec::new(&["*.RS"]);
        let mut flags = PathspecFlags::default();
        flags.ignore_case = true;
        assert!(ps.matches_path("main.rs", &flags));
    }

    #[test]
    fn test_no_glob() {
        let ps = Pathspec::new(&["*.rs"]);
        let mut flags = PathspecFlags::default();
        flags.no_glob = true;
        // With no_glob, *.rs is literal
        assert!(!ps.matches_path("main.rs", &flags));
        assert!(ps.matches_path("*.rs", &flags));
    }

    #[test]
    fn test_match_paths() {
        let ps = Pathspec::new(&["*.rs", "*.toml"]);
        let paths = vec!["src/main.rs", "Cargo.toml", "README.md", "src/lib.rs"];
        let flags = PathspecFlags::default();

        let result = ps.match_paths(&paths, &flags);
        assert_eq!(result.matches.len(), 3);
        assert!(result.matches.contains(&"src/main.rs".to_string()));
        assert!(result.matches.contains(&"Cargo.toml".to_string()));
        assert!(result.matches.contains(&"src/lib.rs".to_string()));
    }

    #[test]
    fn test_match_paths_with_failures() {
        let ps = Pathspec::new(&["*.rs", "*.xyz"]);
        let paths = vec!["src/main.rs"];
        let mut flags = PathspecFlags::default();
        flags.find_failures = true;

        let result = ps.match_paths(&paths, &flags);
        assert_eq!(result.matches.len(), 1);
        assert_eq!(result.failures.len(), 1);
        assert_eq!(result.failures[0], "*.xyz");
    }

    #[test]
    fn test_empty_pathspec_matches_all() {
        let ps = Pathspec::new(&[]);
        let flags = PathspecFlags::default();
        assert!(ps.matches_path("anything.txt", &flags));
    }
}
